File: Problems/Flatten_binary_tree_to_linked_list/fbtl.go
/*
 * +------------------------------------+
 * | Author:        hadi abbasi	        |
 * |                                    |
 * | Link:          github.com/hawwwdi  |
 * +------------------------------------+
 */

package Flatten_binary_tree_to_linked_list

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatter(root *TreeNode) {
	if root != nil && !(root.Left == nil && root.Right == nil) {
		flatter(root.Right)
		flatter(root.Left)
		if tmp := root.Left; tmp != nil {
			for ; tmp.Right != nil; tmp = tmp.Right {
			}
			tmp.Right = root.Right
			root.Right = root.Left
			root.Left = nil
		}
	}
}

File: Problems/Flatten_binary_tree_to_linked_list/fbtl.java
/*
 * +------------------------------------+
 * | Author:        hadi abbasi	        |
 * |                                    |
 * | Link:          github.com/hawwwdi  |
 * +------------------------------------+
 */
//this is my solution(java version) for https://leetcode.com/problems/flatten-binary-tree-to-linked-list/


/* 
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
} */

class Solution {
    public void flatten(TreeNode root) {
        if (root!= null && !(root.left == null && root.right == null)) {
            flatten(root.right);
            flatten(root.left);
            TreeNode tmp = root.left;
            if (tmp != null) {
                for(;tmp.right != null; tmp = tmp.right);
                tmp.right = root.right;
                root.right = root.left;
                root.left = null;
            }
        }
    }
}